saleor/order/views.py

```diff
@@ -107,24 +107,25 @@ def start_payment(request, order, gateway):
                 order.is_fully_paid()
                 or payment.charge_status == ChargeStatus.FULLY_REFUNDED):
             return redirect(order.get_absolute_url())
+
         payment_gateway, gateway_params = get_payment_gateway(payment.gateway)
         client_token = payment_gateway.get_client_token(**gateway_params)
-        form = payment_gateway.get_form_class()
-        form = form(
-            data=request.POST or None, payment=payment,
-            gateway=payment_gateway, gateway_params=gateway_params)
+        form = payment_gateway.get_form_class()(
+            data=request.POST or None, amount=payment.total,
+            currency=payment.currency, gateway_params=gateway_params)
         if form.is_valid():
             try:
-                form.process_payment()
-            except PaymentError as exc:
-                form.add_error(None, exc.message)
+                gateway_process_payment(payment=payment,
```

Comment:
It is better, because the form is then decoupled from the actual logic of the gateway. Why make a form, that should theoretically validate and handle the data, call gateway logic? Now this would be okay if it called Saleor, so Saleor can handle the payment and call gateway methods, but that's impossible if we want to make gateways separate of Saleor. Of course the form could call the gateway methods directly, but that would then split the logic into multiple places and to gain what? It makes it harder to debug and more confusing. It's better to keep the form minimal and return only required stuff for the payment to work (in this case so far it's been the payment token).

Reply:
It makes sense. I'm only afraid that some gateway may not have a way to ```get_payment_token```. But it works so far to the current three payment providers.

Reply:
Yeah, I was worried about that as well, but so far we don't have a case like that and we might never have actually. At the point where this becomes an issue I think payments will be simpler, so implementing such changes to the form and the gateways should be easier as well.